### DISCUSSION TITLE
Why Flows does not have a script block?

### DISCUSSION THREAD

**Discussion by saswatds** [postmanlabs/postman-flows#143] [General]
One of the most requested features in Flows has been to add a Script Block, which is capable of executing javascript. Even though, it seems obvious that scripts should be supported similar to how Requests support running javascript in Test and Pre-Request scripts. 
When designing Flows, we have decided not to make javascript part of the core runtime. It might seem that we are set out to reinvent the wheel, but here are the reasons we think this decision is justified.

1. **Flows is a programming language with is its runtime. Embedding a runtime for another language is messy and unreliable.**
    
    > In simple terms, the abstraction of the *Block in Flows* is not compatible with the idea of a *Function in Javascript*.
    
    The Flows language is a declarative language based on the [Flow-based programming paradigm](https://en.wikipedia.org/wiki/Flow-based_programming). Flows' use of this language paradigm is critical because it can be expressed as a Visual Language. 
    
    Unlike most other programming languages, Flows uses a different computation model called [Actor Model](https://en.wikipedia.org/wiki/Actor_mode) instead of the traditional Turing Machine. The side-effect of this choice is that there is no one-to-one mapping between the language constructs in Flows to any existing imperative languages, like Javascript. The closest match we could find was [Erlang](https://www.erlang.org/)/[Elixir](https://elixir-lang.org/), whose BEAM runtime uses the same actor model. 
    
    In simple terms, the abstraction of the *Block in Flows* is not compatible with the idea of a *Function in Javascript*. Any attempts to introduce compatibly between these two runtimes was causing the language to become unnecessarily complicated in the UI. 
    
    Therefore, we decided not to support javascript in Flows until we standardise the internal [embedding](http://people.cs.aau.dk/~normark/prog3-03/html/notes/languages_themes-embedding-sec.html) APIs.
    
2. **Flows is a strictly typed language; Javascript is not.**
    
    All the cool features you might have got used to in Flows, like schema-based auto-complete, real-time validation, etc. and all the new features that we have planned for the future, depend on Flows being a strictly typed language. Type checking is a foundational pillar for Flows. 
    
    Suppose we were to introduce a script block anywhere with dynamically typed javascript, we immediately become incapable of doing any type-checking. All auto-complete and error-checking features would stop working. The other alternative we have considered was providing a typescript runtime, but unfortunately, browsers do not support typescript natively. Adding typescript would require building a custom typescript runtime and integrating it with the typescript type-system. It would mean a lot of work and prevent us from working on making Flows better.
    
3. **Building a sandbox for javascript is risky and complicated, and keeping up with ECMAScript specifications is a challenge.**
    
    One of our learnings from providing a javascript runtime with Scripts is that maintaining a javascript sandbox is hard. Although browsers offer a reasonable degree of isolation, it makes everyone vulnerable to zero-day exploits. Furthermore, none of the popular javascript runtimes like NodeJS provides any built-in feature to control access to system resources ([deno](https://deno.land/manual/getting_started/permissions) has attempted to solve this with permissions). We make it a moral obligation not to take any decision which puts our users at risk, and for this reason, we have stayed away from letting arbitrary modules in NPM run in Scripts. 
    
    We want to make Flows secure by default and make our users feel safe running flows created by others—all the while encouraging the community to make explicitly safe reusable programs.
    
    Another unspoken aspect of maintaining a javascript runtime is keeping up with the ECMAScript revisions. There is a lot of change happening with javascript because of how popular it is. Hence it is a lot of effort for us to keep up with the changes. 
    

We want to put all our efforts and focus on making Flows better by building very good declarative visual interfaces, which does not need javascript at all. With the [Expression Input](https://github.com/postmanlabs/postman-flows/discussions/124) and the [Abstract Data Editor](https://github.com/postmanlabs/postman-flows/discussions/133) coming in, you'll have the ability to do what you do with JavaScript more intuitively. This method also aligns with our dream of building a collaborative platform for both programmers and non-programmers alike.

**Comment by WilliamWCYoung** [comment]
Thank you for this, it makes a lot of sense and the extra blocks you're adding are really interesting! I can't wait to see how this evolves
